src/Microsoft.AspNetCore.SignalR.Client.Core/HubConnectionBuilder.cs

```diff
@@ -6,68 +6,54 @@
 using System.ComponentModel;
 using Microsoft.AspNetCore.Connections;
 using Microsoft.AspNetCore.SignalR.Internal.Protocol;
+using Microsoft.Extensions.DependencyInjection;
+using Microsoft.Extensions.Logging;
 using Newtonsoft.Json;
 
 namespace Microsoft.AspNetCore.SignalR.Client
 {
     public class HubConnectionBuilder : IHubConnectionBuilder
     {
-        private readonly Dictionary<KeyValuePair<string, Type>, object> _settings = new Dictionary<KeyValuePair<string, Type>, object>();
-        private Func<IConnection> _connectionFactoryDelegate;
-
-        public void ConfigureConnectionFactory(Func<IConnection> connectionFactoryDelegate) =>
-            _connectionFactoryDelegate = connectionFactoryDelegate;
-
-        public void AddSetting<T>(string name, T value)
-        {
-            _settings[new KeyValuePair<string, Type>(name, typeof(T))] = value;
-        }
-
-        public bool TryGetSetting<T>(string name, out T value)
-        {
-            value = default;
-            if (!_settings.TryGetValue(new KeyValuePair<string, Type>(name, typeof(T)), out var setting))
-            {
-                return false;
-            }
-
-            value = (T)setting;
-            return true;
-        }
+        public IServiceCollection Services { get; } = new ServiceCollection();
 
         public HubConnection Build()
         {
-            if (_connectionFactoryDelegate == null)
+            var provider = Services.BuildServiceProvider();
+
+            var connectionFactory = provider.GetService<Func<IConnection>>();
+            if (connectionFactory == null)
             {
-                throw new InvalidOperationException("Cannot create IConnection instance. The connection factory was not configured.");
+                throw new InvalidOperationException("Cannot create HubConnection instance. A connection was not configured.");
             }
 
-            IHubConnectionBuilder builder = this;
-
-            var loggerFactory = builder.GetLoggerFactory();
-            var hubProtocol = builder.GetHubProtocol();
+            var hubProtocol = provider.GetService<IHubProtocol>();
+            var loggerFactory = provider.GetService<ILoggerFactory>();
 
-            return new HubConnection(_connectionFactoryDelegate, hubProtocol ?? new JsonHubProtocol(), loggerFactory);
+            return new HubConnection(connectionFactory, hubProtocol ?? new JsonHubProtocol(), loggerFactory);
```

Comment:
We can use [`ActivatorUtilities.CreateInstance`](https://github.com/aspnet/DependencyInjection/blob/dev/shared/Microsoft.Extensions.ActivatorUtilities.Sources/ActivatorUtilities.cs#L41) to just create `HubConnection` and resolve everything it needs out of DI.

Reply:
But it will be orders of magnitude slower because of reflection. Even using compiled expressions will be slower because they'll be interpreted on Android/iOS.

Reply:
Yea lets avoid that right now.